src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,210 @@
+{
+    // Because a sender can create multiple outputs for us, we first check the outputs vector for an output with
+    // output index 0. If we find it, we remove it from the vector and then iterate over the vector again looking for
+    // an output with index 1, and so on until one of the following happens:
+    //
+    //     1. We have determined all outputs belong to us (the vector is empty)
+    //     2. We have passed over the vector and found no outputs belonging to us
+    //
+
+    bool removed;
+    uint32_t output_index{0};
+    std::vector<uint256> tweaks;
+    do {
+        // We haven't removed anything yet on this pass and if we don't remove anything, we didn't find
+        // any silent payment outputs and should stop checking
+        removed = false;
+        uint256 tweak = ComputeTweak(ecdh_pubkey, output_index);
+        CPubKey computed_sp_pubkey = spend_pubkey.TweakAdd(tweak.data());
+        const XOnlyPubKey& sp_pubkey = XOnlyPubKey{computed_sp_pubkey};
+        //const CKey& silent_payment_priv_key = tweakResult.first;
+        output_pub_keys.erase(std::remove_if(output_pub_keys.begin(), output_pub_keys.end(), [&](auto output_pubkey) {
+            if (sp_pubkey == output_pubkey) {
+                // Since we found an output, we need to increment the output index and check the vector again
+                tweaks.emplace_back(tweak);
+                removed = true;
+                output_index++;
+                // Return true so that this output pubkey is removed the from vector and not checked again
+                return true;
+            }
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && removed);
+    return tweaks;
+}
+
+std::optional<std::pair<uint256, CPubKey>> GetSilentPaymentsTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins)
+{
+
+    // Extract the keys from the inputs
+    // or skip if no valid inputs
+    std::vector<CPubKey> input_pubkeys;
+    std::vector<COutPoint> input_outpoints;
+    for (const CTxIn& txin : vin) {
+        const Coin& coin = coins.at(txin.prevout);
+        Assert(!coin.IsSpent());
+        input_outpoints.emplace_back(txin.prevout);
+
+        std::vector<std::vector<unsigned char>> solutions;
+        TxoutType type = Solver(coin.out.scriptPubKey, solutions);
+        if (type == TxoutType::WITNESS_V1_TAPROOT) {
+            // Check for H point in script path spend
+            if (txin.scriptWitness.stack.size() > 1) {
+                // Check for annex
+                bool has_annex = txin.scriptWitness.stack.back()[0] == ANNEX_TAG;
+                size_t post_annex_size = txin.scriptWitness.stack.size() - (has_annex ? 1 : 0);
+                if (post_annex_size > 1) {
+                    // Actually a script path spend
+                    const std::vector<unsigned char>& control = txin.scriptWitness.stack.back();
```

Comment:
In 644ff77be819a2dea15be0daad10fca8524a18f4 "Implement BIP352: Silent Payments"

Since we aren't modifying `txin.scriptWitness.stack` after checking for the annex, this may not actually be the control block. We need to remove the annex if it was there.

```suggestion
                    const std::vector<unsigned char>& control = txin.scriptWitness.stack.at(post_annex_size - 1);
```

Reply:
Done